### PR TITLE
Add malli schema validation and refactoring

### LINKED ISSUE #10: There is no way to provide a Malli schema
## Problem

There is no way to provide a [metosin/malli](https://github.com/metosin/malli) schema to validate the pulled-data.

## Suggestion

- Add an argument to the `remote-pull` API to provide the `schema` of the expected data.
- Once the data pulled, we validate the data with the schema before sending it to the wire
- We apply this to the sse case as well (might need some refactoring)

### PR DESCRIPTION
**Pull request by skydread1**
Closes #10 

- Refactors pull wrappers and `sse` logics
- Add handler for malli `schema` validation
- Refactors handlers
- Improves testing

### CODE REVIEW

File: src/robertluo/remote_pull/format.clj [hunk 1]
```diff
@@ -45,34 +49,71 @@
...
-        (add-watch a-model ::sse (fn [_ _ _ new-v]
-                                   (if new-v
-                                     (let [data (pull-model new-v pattern opt)]
-                                       (async/put! ch-out data))
-                                     (async/close! ch-out))))
-        (async/put! ch-out data)
-        {:status 200})
-      (throw (ex-info "No pattern" {:req req})))))
+    (let [resp (handler req)]
+      (if-let [schema (get schemas (-> req :body-params :schema))]
```
Comment (robertluo) [inline src/robertluo/remote_pull/format.clj]: `if-let` but just `then` branch, should trigger a warning if you are using clj-kondo.

File: src/robertluo/remote_pull/format.clj [hunk 2]
```diff
@@ -45,34 +49,71 @@
...
+        (if (->> resp :body first (m/validate schema))
+          resp
+          (throw (ex-info "Schema invalid" {:req req})))
```
Comment (robertluo) [inline src/robertluo/remote_pull/format.clj]: Maybe we can put the error reason in error?

File: src/robertluo/remote_pull/format.clj [hunk 3]
```diff
@@ -45,34 +49,71 @@
...
-                                   (if new-v
-                                     (let [data (pull-model new-v pattern opt)]
-                                       (async/put! ch-out data))
-                                     (async/close! ch-out))))
-        (async/put! ch-out data)
-        {:status 200})
-      (throw (ex-info "No pattern" {:req req})))))
+    (let [resp (handler req)]
+      (if-let [schema (get schemas (-> req :body-params :schema))]
+        (if (->> resp :body first (m/validate schema))
```
Comment (robertluo) [inline src/robertluo/remote_pull/format.clj]: Using `m/validator` function will pre-compile the schema.

File: src/robertluo/remote_pull/format.clj (hunk 1 above)
Comment (skydread1) [reply src/robertluo/remote_pull/format.clj]: The else branch is present, it's the `resp`.

File: src/robertluo/remote_pull/format.clj [hunk 4]
```diff
@@ -64,11 +65,15 @@
 (defn with-schema
   [handler schemas]
   (fn [req]
-    (let [resp (handler req)]
+    (let [resp (handler req)
+          data (->> resp :body first)]
       (if-let [schema (get schemas (-> req :body-params :schema))]
-        (if (->> resp :body first (m/validate schema))
+        (if ((m/validator schema) data)
```
Comment (robertluo) [inline src/robertluo/remote_pull/format.clj]:
```clojure
(defn with-schema
   [... schema]
   (let [validator (m/validator schema)]
      ...
```

File: src/robertluo/remote_pull/format.clj [hunk 5]
```diff
@@ -64,11 +65,15 @@
...
           resp
-          (throw (ex-info "Schema invalid" {:req req})))
+          (throw
+           (ex-info "Does not respect schema"
+                    {:req   req
+                     :error (-> schema (m/explain data) me/humanize)})))
```
Comment (robertluo) [inline src/robertluo/remote_pull/format.clj]:
```suggestion
          (throw
            (let [err (m/explain schema data)]
               (ex-info (me/humanize err)
                    {:req   req
                     :error err}))))
```

**Pull request merged**